src/node/txorphanage.cpp

```diff
@@ -83,373 +174,487 @@ class TxOrphanageImpl final : public TxOrphanage {
+
+    // Even though it's possible for MaxPeerAnnouncements to increase within this call to LimitOrphans
+    // (e.g. if a peer's orphans are removed entirely, changing the number of peers), use consistent limits throughout.
+    const auto max_ann{MaxPeerAnnouncements()};
+    const auto max_mem{ReservedPeerUsage()};
+
+    // We have exceeded the global limit(s). Now, identify who is using too much and evict their orphans.
+    // Create a heap of pairs (NodeId, DoS score), sorted by descending DoS score.
+    std::vector<std::pair<NodeId, FeeFrac>> heap_peer_dos;
+    heap_peer_dos.reserve(m_peer_orphanage_info.size());
+    for (const auto& [nodeid, entry] : m_peer_orphanage_info) {
+        // Performance optimization: only consider peers with a DoS score >= 1.
+        if (entry.GetDosScore(max_ann, max_mem) >= FeeFrac{1, 1}) {
+            heap_peer_dos.emplace_back(nodeid, entry.GetDosScore(max_ann, max_mem));
         }
-        // Sweep again 5 minutes after the next entry that expires in order to batch the linear scan.
-        m_next_sweep = nMinExpTime + ORPHAN_TX_EXPIRE_INTERVAL;
-        if (nErased > 0) LogDebug(BCLog::TXPACKAGES, "Erased %d orphan tx due to expiration\n", nErased);
-    }
-    while (m_orphans.size() > DEFAULT_MAX_ORPHAN_TRANSACTIONS)
-    {
-        // Evict a random orphan:
-        size_t randompos = rng.randrange(m_orphan_list.size());
-        EraseTx(m_orphan_list[randompos]->first);
-        ++nEvicted;
     }
-    if (nEvicted > 0) LogDebug(BCLog::TXPACKAGES, "orphanage overflow, removed %u tx\n", nEvicted);
+    static constexpr auto compare_score = [](const auto& left, const auto& right) { return left.second < right.second; };
+    std::make_heap(heap_peer_dos.begin(), heap_peer_dos.end(), compare_score);
+
+    unsigned int num_erased{0};
+    // This outer loop finds the peer with the highest DoS score, which is a fraction of {usage, announcements} used
+    // over the respective allowances. We continue until the orphanage is within global limits. That means some peers
+    // might still have a DoS score > 1 at the end.
+    // Note: if ratios are the same, FeeFrac tiebreaks by denominator. In practice, since the CPU denominator (number of
+    // announcements) is always lower, this means that a peer with only high number of announcements will be targeted
+    // before a peer using a lot of memory, even if they have the same ratios.
+    do {
+        Assume(!heap_peer_dos.empty());
+        // This is a max-heap, so the worst peer is at the front. pop_heap()
+        // moves it to the back, and the next worst peer is moved to the front.
+        std::pop_heap(heap_peer_dos.begin(), heap_peer_dos.end(), compare_score);
+        const auto [worst_peer, dos_score] = std::move(heap_peer_dos.back());
+        heap_peer_dos.pop_back();
+
+        // If needs trim, then at least one peer has a DoS score higher than 1.
+        Assume(dos_score.fee > dos_score.size);
```

Comment:
to get rid of unrelated fee/size terminology, I wonder if a dedicated dos score type would make sense, like e.g.:
```
// Use FeeFrac class to represent the DoS score as fraction,
// with more generic member methods for (de)nominator access
class DosScore : public FeeFrac {
private:
    // prevent direct access of base fraction values
    using FeeFrac::fee, FeeFrac::size;
public:
    using FeeFrac::FeeFrac; // inherit constructors
    int64_t nominator() const { return fee; }
    int32_t denominator() const { return size; }
};
```
(admitedly a bit hacky; can be discussed also in a follow-up PR)

Reply:
Made it >> 1